### PR TITLE
:+1: Add `isParametersOf`

### PR DESCRIPTION
**Pull request by Milly**
## Motivation

I want to validate method arguments with optional parameters.

## Comparison with `isTupleOf()`

- `isTupleOf()`: the length of the tuple is fixed.

```ts
const a: unknown = ["a"];
if (is.TupleOf([is.String, is.Optional(is.Number)] as const)(a)) {
  // don't reach here
  const _: [string, number | undefined] = a;
}
```

- `isParametersOf()`: the tuple is truncated length if trailing predicates are optional.

```ts
const a: unknown = ["a"];
if (is.ParametersOf([is.String, is.Optional(is.Number)] as const)(a)) {
  // reach here
  const _: [string, number?] = a;
}
```

## Naming

If there is a better name, please suggest.

### CODE REVIEW

**Comment by lambdalisue** [comment]
It seems the function is quite resemble with `isTupleOf` with `predElse`.

```ts
import { is } from "https://deno.land/x/unknownutil@$MODULE_VERSION/mod.ts";

const isMyType = is.TupleOf(
  [is.Number, is.String, is.Boolean],
  is.ArrayOf(is.Number),
);
const a: unknown = [0, "a", true, 0, 1, 2];
if (isMyType(a)) {
  // a is narrowed to [number, string, boolean, ...number[]]
  const _: [number, string, boolean, ...number[]] = a;
}
```

What is the difference and advantage compared to the existing one?

**Comment by Milly** [comment]
How can I use `isTupleOf()` with `predElse` to resolve the following `args` type?

It is possible by defining a custom Predicate `isOpts`, but I would like to avoid implementing type guards.

```ts
import { is } from "https://deno.land/x/unknownutil@$MODULE_VERSION/mod.ts";

function f(name: string, optCount?: number, optFlag?: boolean): void {}

const arglist: unknown[] = [
  ["a", 1, true],
  ["a", 1],
  ["a"],
  ["a", true, 1], // Expect skip
  ["a", 1, true, true, true], // Expect skip
];

for (const args of arglist) {
  if (is.TupleOf([is.String] as const, is.ArrayOf(is.UnionOf([is.Number, is.Boolean])))(args)) {
    // args is narrowed to [string, ...(number, boolean)[]]
    f(...args); // deno-ts:Error:2345:Argument of type 'number | boolean' is not assignable to parameter of type 'number | undefined'. Type 'boolean' is not assignable to type 'number'.
  }

  if (is.UnionOf([is.TupleOf([is.String]), is.TupleOf([is.String, is.Number]), is.TupleOf([is.String, is.Number, is.Boolean])])(args)) {
    // args is narrowed to [string] | [string, number] | [string, number, boolean]
    f(...args); // deno-ts:Error:2556:A spread argument must either have a tuple type or be passed to a rest parameter.
  }

  const isOpts = (_x: unknown): _x is [number?, boolean?] => /* nice implements */ true;
  if (is.TupleOf([is.String], isOpts)(args)) {
    // args is narrowed to [string, (number | undefined)?, (boolean | undefined)?]
    f(...args); // No error
  }

  if (is.ParametersOf([is.String, is.OptionalOf(is.Number), is.OptionalOf(is.Boolean)] as const)(args)) {
    // args is narrowed to [string, (number | undefined)?, (boolean | undefined)?]
    f(...args); // No error
  }
}
```

**Comment by lambdalisue** [comment]
Ah got it.

File: is.ts [hunk 1]
```diff
@@ -751,6 +751,129 @@ type IsTupleOfMetadata = {
   args: [Parameters<typeof isTupleOf>[0], Parameters<typeof isTupleOf>[1]?];
 };
 
+/**
+ * Return a type predicate function that returns `true` if the type of `x` is `ParametersOf<T>` or `ParametersOf<T, E>`.
+ *
+ * This is similar to `TupleOf<T>` or `TupleOf<T, E>`, except that the `OptionalOf<P>` at the end of the tuple becomes optional.
```
Comment (lambdalisue) [inline is.ts:757]: I think it is not clear what happen when `OptionalOf` exists in the middle of the tuple with this explanation. It is not must but should be explained a bit more clearly.

For example, like this

```ts
const isMyType = is.ParametersOf([
  is.Number,
  is.OptionalOf(is.String),
  is.Boolean,
  is.OptionalOf(is.Number),
  is.OptionalOf(is.String),
  is.OptionalOf(is.Boolean),
]);

// maybe this?
// type isMyType = Predicate<[number, string | undefined, boolean, number?, string?, boolean?]>;
```

**Review by lambdalisue — APPROVED** [review]
I think it's better to make the description a bit more clear but LGTM.

File: is.ts (hunk 1 above)
Comment (Milly) [reply is.ts]: Fix the doc and examples. How about this?

Comment (Milly) [reply is.ts]: Ah... mistake commit.

**Review by lambdalisue — APPROVED** [review]
💯

**Pull request merged**